src/plugins/clips-executive/clips/domain.clp

```diff
@@ -178,6 +178,260 @@
+  (slot part-of (type SYMBOL)) ; reference to parent formula
+
+  ; a PDDL predicate is either an equality or references a domain-predicate
+  (slot predicate (type SYMBOL)) 
+  (slot equality (type SYMBOL) (allowed-values TRUE FALSE) (default FALSE)) 
+
+  (multislot param-names (type SYMBOL))
+  (multislot param-constants (default (create$)))
+)
+
+(deftemplate grounded-pddl-predicate
+  "A grounded instance of a PDDL predicate. Grounded instances have a value for
+  each parameter slot and are part of a grounded formula."
+  (slot id (type SYMBOL) (default-dynamic (gensym*))) 
+  (slot predicate-id (type SYMBOL)) ; reference to ungrounded base version
+  (slot part-of (type SYMBOL)) ; reference to its grounded parent
+
+  (multislot param-values (default (create$)))
+  (slot grounded (type SYMBOL) (allowed-values TRUE FALSE) (default FALSE))
+  (slot is-satisfied (type SYMBOL) (allowed-values TRUE FALSE) (default FALSE))
+)
+
+(deffunction ground-pddl-predicate
+  "Ground a PDDL predicate based on the given parameter values from an action."
+  (?parent-id ?action-param-names ?action-param-values ?grounded-parent-id)
+
+  (do-for-all-facts ((?predicate pddl-predicate)) 
+        (eq ?parent-id ?predicate:part-of) 
+
+    ;get param names and param values from plan-action and insert into values
+    (bind ?values (create$))
+    (foreach ?param ?predicate:param-names
+      (if (neq (nth$ ?param-index ?predicate:param-constants) nil) then
+        (bind ?values
+          (insert$ ?values ?param-index (nth$ ?param-index ?predicate:param-constants)))
+      else
+        (bind ?action-index (member$ ?param ?action-param-names))
+        (if (not ?action-index) then
+          ; ?param is not in the list of the action parameters
+          (assert (domain-error (error-type unknown-parameter) (error-msg
+            (str-cat "PDDL Predicate " ?predicate:id " has unknown parameter " ?param)))
+          )
+        else
+          (bind ?values
+            (insert$ ?values ?param-index (nth$ ?action-index ?action-param-values)))
+        )
+      )
+    )
+    (assert (grounded-pddl-predicate 
+          (id (sym-cat "grounded-" ?predicate:id "-" (gensym*))) 
+          (predicate-id ?predicate:id) (part-of ?grounded-parent-id) 
+          (param-values ?values))
+    )
+  )
+)
+
+(deffunction ground-pddl-formula
+  "Ground a PDDL formula recursively based on the given values from an action."
+  (?parent-id ?action-param-names ?action-param-values ?grounded-parent-id)
+
+  (bind ?grounded-formulas (create$))
+  (do-for-all-facts ((?formula pddl-formula)) (eq ?parent-id ?formula:part-of) 
+    ;recursively ground subformulas 
+    (bind ?grounded-id (sym-cat "grounded-" ?formula:id "-" (gensym*)))
+    (bind ?grounded-formulas (insert$ ?grounded-formulas 1 ?grounded-id) )
+    (if (eq ?grounded-parent-id "")
+      then 
+      (assert (grounded-pddl-formula (formula-id ?formula:id) 
+                                     (part-of ?grounded-id) 
```

Comment:
Imo it would be better if this was `nil` if there is no parent. If that makes sense, you can scratch the complete if-else.